README.md
<div align="center">

[![Alphanumeric Encoder](logo.gif)](https://www.npmjs.com/package/alphanumeric-encoder)

[![NPM Version](https://img.shields.io/npm/v/alphanumeric-encoder)](https://www.npmjs.com/package/alphanumeric-encoder)
[![NPM Package Size](https://img.shields.io/bundlephobia/min/alphanumeric-encoder)](https://www.npmjs.com/package/alphanumeric-encoder)
[![License](https://img.shields.io/github/license/M-Scott-Lassiter/Alphanumeric-Encoder?color=blue)](https://github.com/M-Scott-Lassiter/Alphanumeric-Encoder/blob/main/LICENSE)

---

[![Tests](https://github.com/M-Scott-Lassiter/Alphanumeric-Encoder/actions/workflows/test.yml/badge.svg)](https://github.com/M-Scott-Lassiter/Alphanumeric-Encoder/actions/workflows/test.yml)
[![Build](https://github.com/M-Scott-Lassiter/Alphanumeric-Encoder/actions/workflows/publish.yml/badge.svg)](https://github.com/M-Scott-Lassiter/Alphanumeric-Encoder/actions/workflows/publish.yml)
[![codecov](https://codecov.io/github/M-Scott-Lassiter/Alphanumeric-Encoder/branch/main/graph/badge.svg?token=MLCXIHQJGA)](https://codecov.io/github/M-Scott-Lassiter/Alphanumeric-Encoder)
[![Dependency status](https://img.shields.io/librariesio/release/npm/alphanumeric-encoder)](https://www.npmjs.com/package/alphanumeric-encoder?activeTab=dependencies)
![Vulnerabilities](https://img.shields.io/snyk/vulnerabilities/npm/alphanumeric-encoder)
[![Open Issues](https://img.shields.io/github/issues/m-scott-lassiter/alphanumeric-encoder/bug)](https://github.com/M-Scott-Lassiter/Alphanumeric-Encoder/labels/bug)

[![Maintained](https://img.shields.io/badge/Maintained%3F-yes-brightgreen.svg)](#envelope-contact)
[![contributions welcome](https://img.shields.io/badge/contributions-welcome-brightgreen.svg?style=flat)](/../../blob/main/CONTRIBUTING.md)

---

</div>

<details open="open">
    <summary><b>Table of Contents</b></summary>

<!-- Note: The toc tags mark autogenerated content. Do not manually modify the content here -->

<!-- toc -->

-   [:book: Getting Started](#book-getting-started)
    -   [Purpose](#purpose)
    -   [Install as an NPM Package](#install-as-an-npm-package)
    -   [Use in Javascript](#use-in-javascript)
-   [:gear: API Documentation](#gear-api-documentation)
-   [:hammer_and_wrench: Node and Operating System Support Policy](#hammer_and_wrench-node-and-operating-system-support-policy)
-   [:clipboard: License and Development](#clipboard-license-and-development)
-   [:envelope: Contact](#envelope-contact)

<!-- tocstop -->

</details>

---

## :book: Getting Started

### Purpose

`alphanumeric-encoder` is a lightweight library with no dependencies. It can [encode](/../../blob/main/API.md#encode) an integer into a letter representation, [decode](/../../blob/main/API.md#decode) the letter back into a number, and [deconstruct](/../../blob/main/API.md#deconstruct) combined strings of letters and numbers into an array of decoded numbers.

This is useful for converting letter indexes (used by people) to numbers (used by computers). Examples include:

-   Spreadsheet columns (e.g. Microsoft Excel's end column is "XFD" which corresponds to 16384)
-   Game boards (e.g. Chess, Battleship) use letters and numbers to identify the grid
-   Geodesy grid reference systems (e.g. the [United States National Grid](https://en.wikipedia.org/wiki/United_States_National_Grid))

### Install as an NPM Package

```javascript
npm install alphanumeric-encoder
```

### Use in Javascript

```javascript
const AlphanumericEncoder = require('alphanumeric-encoder')

const encoder = new AlphanumericEncoder()

console.log(encoder.encode(5)) // 'E'
console.log(encoder.encode(48)) // 'AV'
console.log(encoder.encode(733)) // 'ABE'

console.log(encoder.decode('A')) // 1
console.log(encoder.decode('AC')) // 29
console.log(encoder.decode('ANE')) // 1045

console.log(encoder.deconstruct('C7')) // [3, 7]
console.log(encoder.deconstruct('AC22')) // [29, 22]
console.log(encoder.deconstruct('C3ABC123EFGH456')) // [3, 3, 731, 123, 92126, 456]
```

---

## :gear: API Documentation

See the [API](/../../blob/main/API.md) for more detailed information on how to use the class.

## :hammer_and_wrench: Node and Operating System Support Policy

The test suite has successfully run on all combinations of:

-   [![Node Versions Supported](https://img.shields.io/node/v/alphanumeric-encoder)](https://github.com/nodejs/Release)
-   ![Windows Supported](https://img.shields.io/badge/Windows-0078D6?style=for-the-badge=flat&logo=windows&logoColor=white)
    ![Mac Supported](https://img.shields.io/badge/Mac-000000?style=for-the-badge=flat&logo=apple&logoColor=white)
    ![Linux Supported](https://img.shields.io/badge/Linux-FCC624?style=for-the-badge=flat&logo=linux&logoColor=black)

This project supports [Long-Term Support, Current, and Maintenance](https://github.com/nodejs/Release) versions of node. Once a version reaches end of life, the CI scripts will no longer support them. Odd Node versions will only receive support while in a current status.

Other Node versions and operating systems might support the library, but the tests have not verified other combinations.

---

## :clipboard: License and Development

`alphanumeric-encoder` and all other files in this repository are distributed as free and open-source software under the [MIT License](/../../blob/main/LICENSE), © 2022.

Both [contributions](/../../blob/main/CONTRIBUTING.md) and [bug reports](https://github.com/M-Scott-Lassiter/Alphanumeric-Encoder/issues/new/choose) welcome. See the [change log](/../../blob/main/CHANGELOG.md) for specific details of each release.

Leave a :star2: if you find this project useful!

</div>

## :envelope: Contact

Maintained by M. Scott Lassiter.

[![GitHub Badge Profile](https://img.shields.io/badge/GitHub-100000?style=plastic&logo=github&logoColor=white)](https://github.com/M-Scott-Lassiter)
[![Twitter Badge Profile](https://img.shields.io/badge/Twitter-1DA1F2?style=plastic&logo=twitter&logoColor=white)](https://twitter.com/MScottLassiter)
[![LinkedIn Badge Profile](https://img.shields.io/badge/LinkedIn-0077B5?style=plastic&logo=linkedin&logoColor=white)](https://www.linkedin.com/in/mscottlassiter)
[![Stackoverflow Badge Profile](https://img.shields.io/badge/stackoverflow-orange.svg?longCache=true&style=plastic&logo=stackoverflow&logoColor=white)](https://stackoverflow.com/users/6186333/sandpiper)


CHANGELOG.md
## [1.6.0](https://github.com/M-Scott-Lassiter/Alphanumeric-Encoder/compare/v1.5.2...v1.6.0) (2022-06-11)


### :gift: Feature Changes

* add typescript bindings to the encoder ([e1cd754](https://github.com/M-Scott-Lassiter/Alphanumeric-Encoder/commit/e1cd754472567e61b9f0b420bad3899c988afe0c)), closes [#43](https://github.com/M-Scott-Lassiter/Alphanumeric-Encoder/issues/43)


### :lady_beetle: Bug Fixes

* add typescript definition file to "files" key ([009460c](https://github.com/M-Scott-Lassiter/Alphanumeric-Encoder/commit/009460cb87ebc18b5bbbc6fe90db932232afce1b)), closes [#43](https://github.com/M-Scott-Lassiter/Alphanumeric-Encoder/issues/43)
* correct problems with Typescript definitions and JSDoc comments ([b891650](https://github.com/M-Scott-Lassiter/Alphanumeric-Encoder/commit/b891650b2a0c6891bcb7a265fbd39a06908240cc)), closes [#43](https://github.com/M-Scott-Lassiter/Alphanumeric-Encoder/issues/43)


### :building_construction: Build Changes

* change the semantic-release configuration to a shared config ([c2a19fa](https://github.com/M-Scott-Lassiter/Alphanumeric-Encoder/commit/c2a19fa455444bd07af0968b00d2988369f7680c))
* ignore CHANGELOG.md from linting and formatting ([2e3ca94](https://github.com/M-Scott-Lassiter/Alphanumeric-Encoder/commit/2e3ca941b4343144d551f4e769f1d95b7d64a6ce))
* update commitizen configuration for API ([f40be7f](https://github.com/M-Scott-Lassiter/Alphanumeric-Encoder/commit/f40be7ff983cd7997be370d591b12b04e92e901f))
* update dev dependencies ([150c77b](https://github.com/M-Scott-Lassiter/Alphanumeric-Encoder/commit/150c77bff7baf2e414375b1e2661e23d7f3ea901))

## [1.6.0-beta.3](https://github.com/M-Scott-Lassiter/Alphanumeric-Encoder/compare/v1.6.0-beta.2...v1.6.0-beta.3) (2022-06-04)


### :building_construction: Build Changes

* update commitizen configuration for API ([f40be7f](https://github.com/M-Scott-Lassiter/Alphanumeric-Encoder/commit/f40be7ff983cd7997be370d591b12b04e92e901f))


### :lady_beetle: Bug Fixes

* correct problems with Typescript definitions and JSDoc comments ([b891650](https://github.com/M-Scott-Lassiter/Alphanumeric-Encoder/commit/b891650b2a0c6891bcb7a265fbd39a06908240cc)), closes [#43](https://github.com/M-Scott-Lassiter/Alphanumeric-Encoder/issues/43)

## [1.6.0-beta.2](https://github.com/M-Scott-Lassiter/Alphanumeric-Encoder/compare/v1.6.0-beta.1...v1.6.0-beta.2) (2022-05-20)


### :lady_beetle: Bug Fixes

* add typescript definition file to "files" key ([009460c](https://github.com/M-Scott-Lassiter/Alphanumeric-Encoder/commit/009460cb87ebc18b5bbbc6fe90db932232afce1b)), closes [#43](https://github.com/M-Scott-Lassiter/Alphanumeric-Encoder/issues/43)

## [1.6.0-beta.1](https://github.com/M-Scott-Lassiter/Alphanumeric-Encoder/compare/v1.5.2...v1.6.0-beta.1) (2022-05-20)


### :building_construction: Build Changes

* change the semantic-release configuration to a shared config ([c2a19fa](https://github.com/M-Scott-Lassiter/Alphanumeric-Encoder/commit/c2a19fa455444bd07af0968b00d2988369f7680c))
* ignore CHANGELOG.md from linting and formatting ([2e3ca94](https://github.com/M-Scott-Lassiter/Alphanumeric-Encoder/commit/2e3ca941b4343144d551f4e769f1d95b7d64a6ce))


### :gift: Feature Changes

* add typescript bindings to the encoder ([e1cd754](https://github.com/M-Scott-Lassiter/Alphanumeric-Encoder/commit/e1cd754472567e61b9f0b420bad3899c988afe0c)), closes [#43](https://github.com/M-Scott-Lassiter/Alphanumeric-Encoder/issues/43)

### [1.5.2](https://github.com/M-Scott-Lassiter/Alphanumeric-Encoder/compare/v1.5.1...v1.5.2) (2022-05-14)


### :lady_beetle: Bug Fixes

* correct bug that allowed bad dictionary values to pass through the constructor ([bc6e6c7](https://github.com/M-Scott-Lassiter/Alphanumeric-Encoder/commit/bc6e6c735e12d6eba1827c869430eee3afce65c0)), closes [#41](https://github.com/M-Scott-Lassiter/Alphanumeric-Encoder/issues/41)

### [1.5.1](https://github.com/M-Scott-Lassiter/Alphanumeric-Encoder/compare/v1.5.0...v1.5.1) (2022-05-14)

### :lady_beetle: Bug Fixes

-   improve dictionary type checking ([3f37616](https://github.com/M-Scott-Lassiter/Alphanumeric-Encoder/commit/3f37616376682edeaeed59dfe66a606a61797de4)), closes [#39](https://github.com/M-Scott-Lassiter/Alphanumeric-Encoder/issues/39)

## [1.5.0](https://github.com/M-Scott-Lassiter/Alphanumeric-Encoder/compare/v1.4.0...v1.5.0) (2022-05-03)

### :gift: Feature Changes

-   add the `deconstruct` function to the class ([5df23ec](https://github.com/M-Scott-Lassiter/Alphanumeric-Encoder/commit/5df23ec35290a2cbf7d4c95038af359fd0bf55b5)), closes [#37](https://github.com/M-Scott-Lassiter/Alphanumeric-Encoder/issues/37)

### :dart: Test Changes

-   add test cases for the `deconstruct` function ([1eafa19](https://github.com/M-Scott-Lassiter/Alphanumeric-Encoder/commit/1eafa19b487e7db8697e18fda6847194028eab48)), closes [#37](https://github.com/M-Scott-Lassiter/Alphanumeric-Encoder/issues/37)

## [1.4.0](https://github.com/M-Scott-Lassiter/Alphanumeric-Encoder/compare/v1.3.0...v1.4.0) (2022-05-02)

### :gift: Feature Changes

-   add ability to instantiate the class using an optional config object in the constructor ([8dc230b](https://github.com/M-Scott-Lassiter/Alphanumeric-Encoder/commit/8dc230b164a2689b49d6a7f8b00f51348da8c3f8)), closes [#35](https://github.com/M-Scott-Lassiter/Alphanumeric-Encoder/issues/35)

### :dart: Test Changes

-   add tests verifying the optional config object works as expected ([f7f7de6](https://github.com/M-Scott-Lassiter/Alphanumeric-Encoder/commit/f7f7de6681c4bb3ec1083f346d691d4d36795af3)), closes [#35](https://github.com/M-Scott-Lassiter/Alphanumeric-Encoder/issues/35)

## [1.3.0](https://github.com/M-Scott-Lassiter/Alphanumeric-Encoder/compare/v1.2.0...v1.3.0) (2022-05-01)

### :gift: Feature Changes


.cz-config.js
const encoderMethods = [
    { name: 'dictionary' },
    { name: 'allowLowerCaseDictionary' },
    { name: 'resetDefaultDictionary' },
    { name: 'encode' },
    { name: 'decode' },
    { name: 'deconstruct' }
]

const documentation = [
    { name: 'contributing' },
    { name: 'code of conduct' },
    { name: 'license' },
    { name: 'readme' },
    { name: 'security' }
]

module.exports = {
    types: [
        {
            value: 'api',
            name: 'api:      Non-functional changes to code API documentation that help other developers understand how to use a tool or feature (i.e. intellisense)'
        },
        {
            value: 'build',
            name: 'build:    Changes that affect the build system configuration, package scripts, or dev dependencies (i.e. adds/remove/modify/update)'
        },
        {
            value: 'ci',
            name: 'ci:       Changes to CI configuration files and scripts (e.g. release configs, YAML scripts)'
        },
        { value: 'docs', name: 'docs:     Documentation only changes' },
        { value: 'feat', name: 'feat:     Adds a new feature' },
        {
            value: 'fix',
            name: 'fix:      Fixes a bug in an existing feature. Also used for non-dev dependency updates.'
        },
        {
            value: 'perf',
            name: 'perf:     A code change that improves performance'
        },
        {
            value: 'refactor',
            name: 'refactor: A code change that neither fixes a bug nor adds a feature'
        },
        { value: 'revert', name: 'revert:   Revert to a commit' },
        { value: 'test', name: 'test:     Add missing tests or correct existing tests' }
    ],

    scopes: [...encoderMethods, ...documentation],

    allowTicketNumber: false,
    isTicketNumberRequired: false,
    ticketNumberPrefix: 'TICKET-',
    ticketNumberRegExp: '\\d{1,5}',

    // it needs to match the value for field type. Eg.: 'fix'
    scopeOverrides: {
        api: encoderMethods,
        ci: [{ name: 'publish' }, { name: 'test' }, { name: 'hooks' }],
        docs: documentation,
        fix: encoderMethods,
        perf: encoderMethods,
        refactor: encoderMethods,
        test: encoderMethods
    },

    // override the messages, defaults are as follows
    messages: {
        type: "Select the type of change that you're committing:",
        scope: '\nDenote the SCOPE of this change (optional):',
        // used if allowCustomScopes is true
        customScope: 'Denote the SCOPE of this change:',
        subject: 'Write a SHORT, IMPERATIVE tense description of the change:\n',
        body: 'Provide a LONGER description of the change (optional). Use "|" to break new line:\n',
        breaking: 'List any BREAKING CHANGES (optional):\n',
        footer: 'List any ISSUES CLOSED by this change (optional). E.g.: #31, #34:\n',
        confirmCommit: 'Are you sure you want to proceed with the commit above?'
    },

    allowCustomScopes: true,
    allowBreakingChanges: ['feat', 'fix'],
    // skip any questions you want
    // skipQuestions: ['body'],

    // limit subject length
    subjectLimit: 100,
    // breaklineChar: '|', // It is supported for fields body and footer.
    footerPrefix: 'Resolves:'
    // askForBreakingChangeFirst : true, // default is false
}


release.config.js
module.exports = {
    branches: ['main', 'next'],
    plugins: [
        [
            '@semantic-release/commit-analyzer',
            {
                preset: 'angular',
                releaseRules: [
                    { type: 'build', release: 'patch' },
                    { scope: 'api', release: 'patch' }
                ]
            }
        ],
        [
            '@semantic-release/release-notes-generator',
            {
                preset: 'conventionalcommits',
                presetConfig: {
                    types: [
                        {
                            type: 'build',
                            section: ':building_construction: Build Changes',
                            hidden: false
                        },
                        { type: 'ci', hidden: true },
                        { type: 'docs', hidden: true },
                        { type: 'feat', section: ':gift: Feature Changes', hidden: false },
                        { type: 'fix', section: ':lady_beetle: Bug Fixes', hidden: false },
                        {
                            type: 'perf',
                            section: ':fast_forward: Performance Improvements',
                            hidden: false
                        },
                        { type: 'refactor', hidden: true },
                        { type: 'test', section: ':dart: Test Changes', hidden: false }
                    ]
                },
                writerOpts: {
                    commitsSort: ['subject', 'scope']
                }
            }
        ],
        [
            '@semantic-release/changelog',
            {
                changelogFile: 'CHANGELOG.md'
            }
        ],
        '@semantic-release/npm',
        ['@semantic-release/github'],
        [
            '@semantic-release/git',
            {
                assets: [
                    'index.js',
                    'README.md',
                    'API.md',
                    'CHANGELOG.md',
                    'CONTRIBUTING.md',
                    'LICENSE',
                    'package.json',
                    'package-lock.json'
                ],
                // eslint-disable-next-line no-template-curly-in-string
                message: 'docs: ${nextRelease.version} [skip ci]\n\n${nextRelease.notes}' // These variables are used in the CD script by semantic-release
            }
        ]
    ],
    preset: 'angular'
}


index.js
// @ts-check

/**
 * A class for encoding and decoding base 10 integers to a custom alphanumeric base representation.
 * @param {object} [configOptions] Optional object defining initial settings for the class
 * @param {boolean} [configOptions.allowLowerCaseDictionary=false] Whether or not to allow lower case letters in the dictionary
 * @param {string} [configOptions.dictionary='ABCDEFGHIJKLMNOPQRSTUVWXYZ'] Starting dictionary to use. Must contain only letters or numbers. Characters cannot be repeated.
 * If `allowLowerCaseDictionary = true`, then lower case letters are not considered the same as upper case. (e.g. 'ABCabc' has 6 unique characters.)
 * @example
 * // Import into a project
 * const AlphanumericEncoder = require('alphanumeric-encoder')
 *
 * const encoder = new AlphanumericEncoder()
 *
 * @example
 * // Import into a project
 * const AlphanumericEncoder = require('alphanumeric-encoder')
 *
 * const encoder = new AlphanumericEncoder({allowLowerCaseDictionary: true, dictionary: "abcdEFGH"})
 *
 * @example
 * // Import into a project
 * const AlphanumericEncoder = require('alphanumeric-encoder')
 *
 * const configOptions = {allowLowerCaseDictionary: true, dictionary: "abcdEFGH"}
 * const encoder = new AlphanumericEncoder(configOptions)
 */
class AlphanumericEncoder {
    constructor(
        configOptions = {
            allowLowerCaseDictionary: false,
            dictionary: 'ABCDEFGHIJKLMNOPQRSTUVWXYZ'
        }
    ) {
        /**
         * @private
         * @type {string} Internal value used to initialize and reset the dictionary
         * @default The English alphabet, all capitalized, in order
         * */
        this._defaultDictionary = 'ABCDEFGHIJKLMNOPQRSTUVWXYZ'
        /**
         * @private
         * @type {string} Internal value of the dictionary to encode/decode against
         */
        this._dictionaryInUse = ''
        /**
         * @private
         * @type {boolean} Internal value that tracks whether or not the dictionary setter should allow lower case letters or not
         * */
        this._allowLowerCaseDictionary = false

        this.resetDefaultDictionary()

        // Process the options. If the user included any, then the if statements will evaluate truthy and try to
        //  set the appropriate values.
        if ('allowLowerCaseDictionary' in configOptions) {
            this.allowLowerCaseDictionary = configOptions.allowLowerCaseDictionary
        }
        if ('dictionary' in configOptions) {
            this.dictionary = configOptions.dictionary
        }
    }

    /**
     * Returns or sets the current dictionary.
     *
     * @param {string} newDictionary (If setting) String of unique letters and numbers, in order, for the new dictionary
     * @throws {RangeError} if setting dictionary to `null`, `undefined` or empty string (i.e. `''`)
     * @throws {RangeError} if `newDictionary` contains a non-alphanumeric character
     * @throws {RangeError} if `newDictionary` has a repeating character
     * @returns {string} (If used as getter) The current dictionary in use
     * @default `ABCDEFGHIJKLMNOPQRSTUVWXYZ`
     *
     * @example
     * const encoder = new AlphanumericEncoder()
     *
     * console.log(encoder.dictionary)  // 'ABCDEFGHIJKLMNOPQRSTUVWXYZ'
     *
     * encoder.dictionary = 'ABCD'
     * console.log(encoder.dictionary)  // 'ABCD'
     *
     * encoder.dictionary = 'ABCDA' // Throws error because the letter 'A' is repeated
     */
    set dictionary(newDictionary) {
        // Check for empty or wrong type dictionaries
        if (
            typeof newDictionary !== 'string' ||
            newDictionary.length === 0 ||
            // eslint-disable-next-line no-self-compare
            newDictionary !== newDictionary // This verifies it wasn't passed NaN
        ) {
            throw new RangeError(
                'The dictionary cannot be null, undefined, boolean, NaN, or an empty string.'
            )
        }

        // Check for invalid characters. Using a regular expression, make sure only letters and numbers are allowed.
        const regExPattern = /^[a-z0-9]+$/i
        if (!regExPattern.test(newDictionary)) {
            throw new RangeError('All characters in the dictionary must be alphanumeric.')
        }

        // Convert to upper case only unless allowing for lower case letters
        let formattedDictionary = newDictionary
        if (!this._allowLowerCaseDictionary) {
            formattedDictionary = newDictionary.toUpperCase()
        }

        // Verify each character is only used one time within the dictionary.
        for (let i = 0; i < formattedDictionary.length; i++) {
            if (
                formattedDictionary.indexOf(formattedDictionary[i]) !==
                formattedDictionary.lastIndexOf(formattedDictionary[i])
            ) {
                throw new RangeError(
                    `The dictionary in use has at least one repeating symbol: ${formattedDictionary[i]}`
                )
            }
        }

        // Validation is complete. Update the internal property.
        this._dictionaryInUse = formattedDictionary
    }

    get dictionary() {
        return this._dictionaryInUse
    }

    /**
     * Returns or sets a boolean value that determines whether the dictionary will allow lower case letters or not.
     *
     * @param {boolean} isAllowed (If setting). Accept truthy or falsy statements.
     * @returns {boolean} (If used as getter)
     * @default false
     * @example
     * const encoder = new AlphanumericEncoder()
     * encoder.dictionary = 'abcdefg' // Default for `allowLowerCaseDictionary` is false
     * console.log(encoder.dictionary) // 'ABCDEFG'
     *
     * @example
     * const encoder = new AlphanumericEncoder()
     * encoder.allowLowerCaseDictionary = true
     * encoder.dictionary = 'ABCDefg'
     * console.log(encoder.dictionary) // 'ABCDefg'
     */
    set allowLowerCaseDictionary(isAllowed) {
        this._allowLowerCaseDictionary = !!isAllowed // The double !! converts truthy or falsy values to true or false, respectively
    }

    get allowLowerCaseDictionary() {
        return this._allowLowerCaseDictionary
    }

    /**
     * Reset the dictionary in use to the default.
     * @default `ABCDEFGHIJKLMNOPQRSTUVWXYZ`
     * @returns {void}
     * @example
     * const encoder = new AlphanumericEncoder()
     * console.log(encoder.dictionary) // 'ABCDEFGHIJKLMNOPQRSTUVWXYZ'
     * encoder.dictionary = 'ABCD'
     * console.log(encoder.dictionary) // 'ABCD'
     * encoder.resetDefaultDictionary()
     * console.log(encoder.dictionary) // 'ABCDEFGHIJKLMNOPQRSTUVWXYZ'
     */
    resetDefaultDictionary() {
        this._dictionaryInUse = this._defaultDictionary
    }

    /**
     * Takes any number and converts it into a base (dictionary length) letter combo.
     *
     * @param {number} integerToEncode Base 10 integer. If passed a non-integer number, decimal values are truncated.
     * Passing zero, negative numbers, or non-numbers will return `undefined`.
     * @throws {RangeError} if `integerToEncode` exceeds the maximum safe integer for Javascript (`2^53 - 1 = 9007199254740991`).
     * @returns {string|undefined} Dictionary encoded value. Returns undefined of passed negative numbers or NaN
     *
     * @example
     * const encoder = new AlphanumericEncoder()
     * console.log(encoder.encode(5)) // 'E'
     * console.log(encoder.encode(48)) // 'AV'
     * console.log(encoder.encode(733)) // 'ABE'
     *
     * @example
     * encoder.dictionary = 'ABCD'
     * console.log(encoder.encode(5)) // 'AA'
     * console.log(encoder.encode(48)) // 'BCD'
     * console.log(encoder.encode(733)) // 'BCACA'
     *
     * @example
     * encoder.dictionary = 'DCBA'
     * console.log(encoder.encode(5)) // 'DD'
     * console.log(encoder.encode(48)) // 'CBA'
     * console.log(encoder.encode(733)) // 'CBDBD'
     *
     * @example
     * encoder.dictionary = 'ABC123'
     * console.log(encoder.encode(5)) // '2'
     * console.log(encoder.encode(48)) // 'AA3'
     * console.log(encoder.encode(733)) // 'CBBA'
     *
     * @example
     * console.log(encoder.encode('A')) // undefined
     * console.log(encoder.encode(null)) // undefined
     * console.log(encoder.encode(undefined)) // undefined
     */
    encode(integerToEncode) {
        if (Number.isNaN(integerToEncode) || integerToEncode < 0) {
            return undefined
        }
        if (integerToEncode > Number.MAX_SAFE_INTEGER) {
            throw new RangeError(
                'The encoding value is greater than the maximum safe integer for Javascript.'
            )
        }

        /**
         * Takes a letter between 0 and max letter length and returns the corresponding letter
         *
         * @private
         * @param {number} num The index to search for
         * @param {string} dictionary The dictionary used to search
         * @returns {string} The letter corresponding to this number index
         */
        function numToLetter(num, dictionary) {
            if (num === 0) {
                return undefined
            }
            return dictionary.slice(num - 1, num)
        }

        const baseNumber = Math.abs(Math.floor(integerToEncode))

        let index = baseNumber % this.dictionary.length
        let quotient = baseNumber / this.dictionary.length

        if (baseNumber <= this.dictionary.length) {
            // Number is within single digit bounds of our encoding letter alphabet
            return numToLetter(baseNumber, this.dictionary)
        }

        // By reaching this point in the code, this number was bigger than our dictionary. Recursively perform this function until complete
        if (index === 0) {
            quotient -= 1
        } // Accounts for the edge case of the last letter in the dictionary string
        const result = this.encode(quotient)

        if (index === 0) {
            index = this.dictionary.length
        } // Accounts for the edge case of the final letter; avoids getting an empty string

        return result + numToLetter(index, this.dictionary)
    }

    /**
     * Takes any string and converts it into a base 10 integer based on the defined dictionary.
     *
     * @param {string} stringToDecode If passed a non-integer number, decimal values are truncated.
     * Passing an empty string, `null`, or `undefined` will return `undefined`.
     * @throws {RangeError} if the decoded integer exceeds the maximum safe integer for Javascript (`2^53 - 1 = 9007199254740991`).
     * @returns {number} Positive integer representation. If one of the characters is not present in the dictionary, it will return `undefined`.
     * @example
     * const encoder = new AlphanumericEncoder()
     * console.log(encoder.decode('A')) // 1
     * console.log(encoder.decode('AC')) // 29
     * console.log(encoder.decode('ANE')) // 1045
     *
     * @example
     * console.log(encoder.decode('a')) // undefined
     * console.log(encoder.decode(123)) // undefined
     * console.log(encoder.decode('A?')) // undefined
     * console.log(encoder.decode(null)) // undefined
     * console.log(encoder.decode(undefined)) // undefined
     *
     * @example
     * encoder.dictionary = 'ABCD'
     * console.log(encoder.decode('A')) // 1
     * console.log(encoder.decode('AC')) // 7
     * console.log(encoder.decode('ADBAC')) // 551
     * console.log(encoder.decode('ANE')) // undefined
     */
    decode(stringToDecode) {
        // Takes any number encoded with the provided encode dictionary

        const safeEncoded = String(stringToDecode) // += '' // Force any numbers passed into a string format to get proper behavior below

        if (safeEncoded === '' || safeEncoded === null || safeEncoded === undefined) {
            return undefined
        }

        let result = 0
        let index = 0

        for (let i = 1; i <= safeEncoded.length; i++) {
            index = this.dictionary.search(safeEncoded.slice(i - 1, i)) + 1
            if (index === 0) {
                return undefined
            } // Attempted to find a letter that wasn't encoded in the dictionary
            result += index * this.dictionary.length ** (safeEncoded.length - i)
        }

        if (result > Number.MAX_SAFE_INTEGER) {
            throw new RangeError(
                'The decoded value is greater than the maximum safe integer for Javascript.'
            )
        }

        return result
    }

    /**
     * Takes any string of letters and numbers and deconstructs it into an array of base 10 integers based on the defined dictionary.
     *
     * @param {string} stringToDeconstruct A string of letters and numbers (e.g. `'A7'`, `'AC22'`, `'7C10F'`)
     * @throws {Error} if the dictionary contains a number as this function would be unable to differentiate between where a number and dictionary value.
     * @returns {(number|undefined)[]|undefined} An array of numbers. Characters not present in the dictionary are treated as letters and return `undefined` for that array value.
     * Passing an empty string (`''`), `null`, or `undefined` will return `undefined` for the whole function.
     * @example
     * const encoder = new AlphanumericEncoder()
     * console.log(encoder.deconstruct('A')) // [1]
     * console.log(encoder.deconstruct('AC22')) // [29, 22]
     * console.log(encoder.deconstruct('C3ABC123EFGH456')) // [3, 3, 731, 123, 92126, 456]
     * console.log(encoder.deconstruct('A1aB2B')) // [1, 1, undefined, 2, 2]
     * console.log(encoder.deconstruct('7AC!23A1%')) // [7, undefined, 23, 1, 1, undefined]
     * console.log(encoder.deconstruct('')) // undefined
     *
     */
    deconstruct(stringToDeconstruct) {
        // The dictionary cannot contain numbers, or else the deconstruct function cannot distinguish where
        //  one code begins and another ends.
        if (this.dictionary.match(/[0-9]/)) {
            throw new Error('Cannot deconstruct if the dictionary contains numbers.')
        }

        // Passing falsy values should return undefined
        if (
            stringToDeconstruct === null ||
            stringToDeconstruct === undefined ||
            String(stringToDeconstruct).length === 0
        ) {
            return undefined
        }

        const safeString = String(stringToDeconstruct) // Force argument to string to process number arguments and prevent slice from throwing an error
        /**
         * @private
         * @type {number[]} Array of numbers after splitting a string.
         * Elements include (in order) the number or decoded number from a letter.
         */
        const deconstructedArray = []
        let character = ''
        let componentPart = safeString.slice(0, 1) // Initialize with the first character (which has been guranteed present by above guard functions)

        // A helper function to push the final component into the array that gets returned. Numbers get added as is, strings get decoded.
        const addDecodedElement = (/** @type {string} */ componentString) => {
            if (componentString.match(/[0-9]/)) {
                deconstructedArray.push(Number.parseInt(componentString, 10)) // Numbers
            } else {
                deconstructedArray.push(this.decode(componentString)) // Letters
            }
        }

        // If more than one character in safeString, loop through each subsequent character. Once the next character is not
        // the same type as the previous group (i.e. flips from letter to number, or vice versa), add the character group to
        // deconstructedArray, reset, and move to the next.
        for (let i = 2; i <= safeString.length; i++) {
            character = safeString.slice(i - 1, i)

            // Parse using a RegExp looking for numbers. The !! converts this to either true/false.
            if (!!character.match(/[0-9]/) === !!componentPart.match(/[0-9]/)) {
                // Same type, concatenate and keep going
                componentPart += character
            } else {
                // Flipped types, add to array and reset
                addDecodedElement(componentPart)
                componentPart = character
            }
        }

        // Add the final component part (for single character stringToDeconstruct, this will be the only part)
        addDecodedElement(componentPart)

        return deconstructedArray
    }
}

module.exports = AlphanumericEncoder
